### PR TITLE
feat: add DAG class

### PR DESCRIPTION
**Pull request by peternhale**
@mdonnalley @RodEsp just wanted to get the graph impl reviewed as I start the integration with this work and config.

DO NOT MERGE

### CODE REVIEW

**Review by mdonnalley — COMMENTED** [review]
Looks great so far - awesome work!

File: src/config/graph.ts [hunk 1]
```diff
@@ -0,0 +1,139 @@
...
+  setNode(name: string, node: N): void {
+    this._nodes.set(name, node)
+  }
+
+  setEdge(source: N, sink: N): EdgeClass<N, E> {
+    this.setNode(source.id, source)
+    this.setNode(sink.id, sink)
+
+    const edge = new EdgeClass<N, E>(source, sink)
+    if (this._edges.has(edge)) {
```
Comment (amphro) [inline src/config/graph.ts:49]: Will this every return true? `edge` is a new edge instance which won't be in the set. I think you want something like `this._edges.find(edge => edge.source.id === source.id && edge.sink.id === sink.id)`?

Comment (peternhale) [reply src/config/graph.ts:49]: I will check

Comment (peternhale) [reply src/config/graph.ts:49]: Nope, always creates a new edge. Nice catch.

**Pull request closed without merging**